src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -54,110 +54,74 @@ def _use_object(self, object_type: UseObjectType, name: str):
             handle_unclassified_error(err, f"Failed to use {object_type} {name}.")
 
     @contextmanager
-    def _use_warehouse_optional(self, new_wh: str | None):
+    def _use_object_optional(self, object_type: UseObjectType, name: str | None):
         """
-        Switches to a different warehouse for a while, then switches back.
-        This is a no-op if the requested warehouse is already active or if no warehouse is passed in.
-        @param new_wh: Name of the warehouse to use. If not a valid Snowflake identifier, will be converted before use.
+        Call sql to use snowflake object with error handling
+        @param object_type: ObjectType, type of snowflake object to use
+        @param name: object name, will be cast to a valid snowflake identifier.
         """
-        if new_wh is None:
+        if name is None:
             yield
             return
 
-        valid_wh_name = to_identifier(new_wh)
-
         try:
-            wh_result = self._sql_executor.execute_query(
-                "select current_warehouse()"
+            current_obj_result_row = self._sql_executor.execute_query(
+                f"select current_{object_type}()"
             ).fetchone()
         except Exception as err:
-            return handle_unclassified_error(err, "Failed to select current warehouse.")
+            return handle_unclassified_error(
+                err, f"Failed to select current {object_type}."
+            )
 
-        # If user has an assigned default warehouse, prev_wh will contain a value even if the warehouse is suspended.
         try:
-            prev_wh = wh_result[0]
-        except:
-            prev_wh = None
-        # new_wh is not None, and should already be a valid identifier, no additional check is performed here.
-        is_different_wh = valid_wh_name != prev_wh
-        if is_different_wh:
-            self._log.debug(f"Using warehouse: {valid_wh_name}")
-            self._use_object(UseObjectType.WAREHOUSE, valid_wh_name)
+            prev_obj = current_obj_result_row[0]
+        except IndexError:
+            prev_obj = None
+
+        if prev_obj is not None and same_identifier(prev_obj, name):
+            yield
+            return
```

Comment:
the body here is the same, but the logic was inverted here to return early since I think it makes it cleaner )and it matches the logic at the top of the method)